docs/proposals/bulward-integration.md

```diff
@@ -0,0 +1,206 @@
+For example, both `couchdb-consumer` and `redisdb-operator` will all have access to both `couchdb` and `redisdb` resources.
+A possible workaround would be using two `Account`s for `couchdb-consumer` and `redisdb-operator`, but that just misses the point
+to have the option to make an `Account` be Provider and Tenant at the same time.
+to solve this issue, we can use `Organization` to represent `Account`:
+```yaml
+apiVersion: apiserver.bulward.io/v1alpha1
+kind: Organization
+metadata:
+  name: company-a
+spec:
+  metadata:
+    displayName: The A Company
+    description: In 1972, a crack commando unit was sent to prison by a military court...
+  owners:
+  - kind: User
+    name: company-owner
+    apiGroup: rbac.authorization.k8s.io
+  - kind: User
+    name: couchdb-consumer
+    apiGroup: rbac.authorization.k8s.io
+  - kind: User
+    name: redisdb-operator
+    apiGroup: rbac.authorization.k8s.io
+```
+And then we can have dedicated `Project` for both `couchdb` and `redisdb`:
+```yaml
+apiVersion: apiserver.bulward.io/v1alpha1
+kind: Project
+metadata:
+  name: couchdb
+  namespace: organization-a
+spec:
+  owners:
+  - kind: User
+    name: owner
+    apiGroup: rbac.authorization.k8s.io
+  - kind: User
+    name: couchdb-consumer
+    apiGroup: rbac.authorization.k8s.io
+---
+apiVersion: apiserver.bulward.io/v1alpha1
+kind: Project
+metadata:
+  name: redisdb
+  namespace: organization-a
+spec:
+  owners:
+  - kind: User
+    name: owner
+    apiGroup: rbac.authorization.k8s.io
+  - kind: User
+    name: redisdb-operator
+    apiGroup: rbac.authorization.k8s.io
+```
+In this way, we can easily grant different permissions for different team, and different projects. For example, in bulward,
+we can use `ProjectRoleTemplate` and `ProjectRole` to set up permission for a (group of) `Project`.
+
+Overall:
+- `Account` should be mapped to `Organzation` in a way, not `Project`.
+- `Organization` can be `Provider` and `Tenant` at the same time, but `Project` under
+`Organization` should be defined as Provider project or Tenant project.
+
+## Implementation
+### Deployment
+Every bulward release can contain a manifest yaml file, just like cert-manager, we can install it in the similar way:
+```bash
+kubectl apply -f https://github.com/kubermatic/bulward/releases/download/${BULWARD_VERSION}/bulward.yaml
+kubectl wait --for=condition=available deployment/bulward-controller-manager -n bulward-system --timeout=120s
+kubectl wait --for=condition=available deployment/bulward-apiserver -n bulward-system --timeout=120s
+```
+### API Server
+#### Organization Endpoint
+We have an endpoint for `Account`, we can reuse it for `Organization`, also, in Bulward, only `Organization` object is visible
+for organization member (this is implemented in Bulward api extension server), we can also drop the filtering implementation in currently account endpoint.
```

Comment:
Yes and no. 

Correct me if I'm wrong, in our current API server implementation we use the same REST config for our k8s client, without any user impersonation? Thus when we query `apiserver.bulward.io` we would have same visibility as API server service account. 

Thus we either need to query `storage.bulward.io` and do API server filtering ( a duplicate work, and bad approach ). Or get the user REST config impersonation working per API server backend request.

Reply:
> Correct me if I'm wrong, in our current API server implementation we use the same REST config for our k8s client, without any user impersonation?

Yes, that' right. 
In our `Account` endpoint, we have a filtering layer now, we use `FieldIndexer` to fetch the `Account` objects, which has user name in the `Account.Spec.Subjects`.

When we integrate KubeCarrier with Bulward, as you mentioned, I would like to use an impersonation client, so we can cut the current filtering layer in API server, and offload that to Bulward completely.

Reply:
we could also use the same impersonation to replace or complement our authZ layer as well.